File: javascript/queen-attack/queen-attack.js
export class QueenAttack {

    constructor(
        {white=[0, 3], black=[7, 3]} 
        = {white:[0, 3], black:[7, 3]}) {

        if (white[0] === black[0] && white[1] === black[1])
            throw new Error('Queens cannot share the same space');
        
        this.white = white;
        this.black = black;
        this.setupBoard();
    }

    setupBoard() {
        this.board =    
        [   ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_'],
            ['_', '_', '_', '_', '_', '_', '_', '_']    ];

        this.board[this.white[0]][this.white[1]] = 'W';
        this.board[this.black[0]][this.black[1]] = 'B';
    }

    toString() {
        return this.board.map(arr => arr.join(' ')).join('\n') + '\n';
    }

    canAttack() {
        const { white, black } = this;
        return white[0] === black[0]  // Same row
            || white[1] === black[1]  // Same column
            // Same nw-se diagonal
            || white[0] - white[1] === black[0] - black[1]
            // Same ne-sw diagonal
            || white[0] + white[1] === black[0] + black[1];
    }
}

File: cpp/queen-attack/queen_attack.cpp
#include "queen_attack.h"
#include <stdexcept>

namespace queen_attack {

	chess_board::chess_board() :
		chess_board(std::make_pair(0, 3), std::make_pair(7, 3)) {}
	
	chess_board::chess_board(std::pair<int, int> white, std::pair<int, int> black)
	{
		if (white == black)
			throw std::domain_error("Queens may not occupy the same square.");

		_white = white;
		_black = black;

		_board = {
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n"
			"_ _ _ _ _ _ _ _\n" };

		_board[(__int64)_white.first * 16 + (__int64)_white.second * 2] = 'W';
		_board[(__int64)_black.first * 16 + (__int64)_black.second * 2] = 'B';

	}

	const std::pair<int, int> chess_board::white() const { return _white; }
	const std::pair<int, int> chess_board::black() const { return _black; }

	bool chess_board::can_attack() const
	{
		return _white.first == _black.first // Same row
			|| _white.second == _black.second // Same column
			// Same NW-SE diagonal
			|| _white.first - _black.first == _white.second - _black.second
			// Same NE-SW diagonal
			|| _white.first + _white.second == _black.first + _black.second;
	}


}  // namespace queen_attack
